### ISSUE TITLE
In cdirect.c, divide_good_rects(), are K1 and im really needed

### ISSUE THREAD

**Issue by bhp1** [stevengj/nlopt#82]
In a convex hull, the slope before a point should always be less than the slope after a point.  In cdirect/cdirect.c, divide_good_rects() calculates both the slope before, K1, and the slope after, K2, then computes:
`K = MAX(K1, K2);`
Won't this always result in K == K2, making K1 and im unnecessary?

Further, the definition of a Potentially Optimal Hyperrectangle requires K > 0.  The test involving magic_eps will ensure K >= 0.  However, should divide_good_rects() enforce K > 0?  (Essentially, this means not dividing any rectangles smaller than the largest rectangle that has f == fmin.)

**Comment by stevengj** [comment]
Good point; I think K1 is only used if the ip loop is empty, i.e. if `i == nhull`.

**Comment by stevengj** [comment]
Enforcing `K > 0` seems reasonable, but it's been a while since I thought about this algorithm and I would have to think about it for a while.  In practice, however, it seems unlikely to matter, since you're unlikely to get multiple rectangles with `f == fmin`, no?

**Comment by bhp1** [comment]
I admit the chances of `f == fmin` in rectangles of different diameter look slim.  However, I think enforcing `K > 0` rather than just `K >= 0` has merit for several reasons:
1. It's in the definition of POH.
2. Ultimately, the condition "`eps` arbitrarily small but greater than zero" reduces to `K > 0, K != 0`, so you can avoid all the debates about, "How small should `eps` be?"
3. It ensures that the hull always includes fmin.  You avoid the sensitivity to additive scaling that Daniel Finkel pointed out.
4. Turns out, it can be implemented in the convex hull pruning process.  Rather than the lower half of the convex hull, return the lower right quarter.  `divide_good_rects` doesn't have to do any more pruning.

BTW, I want to thank you for posting your cdirect implementation of DIRECT.  I've translated cdirect into Java, in the style of [Apache Commons Math](http://commons.apache.org/proper/commons-math/userguide/optimization.html#a12.4_Direct_Methods), for [the WIDesigner project](/edwardkort/WWIDesigner).  cdirect was far easier to work with than the Fortran and MatLab implementations also available, and helped me understand how DIRECT really works.  (I last worked with Fortran around the time Fortran 77 came out.)  I don't think I could have done it without your work.  I'll let you know when I've posted the Java version to WIDesigner.

**Comment by bhp1** [comment]
You can find the Java versions of cdirect at [https://github.com/edwardkort/WWIDesigner/tree/optimizer/WWIDesigner/src/main/com/wwidesigner/math], split into separate versions for the original DIRECT algorithm (DIRECTOptimizer.java), Gablonsky's DIRECT-L (DIRECT_L_Optimizer.java), and a variant of the cdirect variant that divides rectangles other than hypercubes on only one of the long sides (DIRECT1Optimizer.java).  I have found the last of these generally gives much faster results.  This is illustrated in the JUnit tests of each on various standard functions in [https://github.com/edwardkort/WWIDesigner/tree/optimizer/WWIDesigner/src/test/com/wwidesigner/math].

Comments are welcome.  Thanks again for making this possible.

**Comment by stevengj** [comment]
If you ever feel like submitting a PR for this, that would be great.
